### PR TITLE
Git over HTTP

### PR DESCRIPTION
**Pull request by titanous**
This changes gitreceived to use the Smart HTTP git protocol instead of SSH, which removes the need for users to manage SSH keys.

The CLI now configures two global git options that pin the CA certificate for the git remote and use a credential helper built into the CLI to get the password. In the future this helper will be used to support per-user auth, etc.

Port 2222 is no longer used and is removed from all configs/docs.

The CLI cluster config is simplified, but remains temporarily backwards-compatible with SSH gitreceive, we'll remove the backwards compatibility when few users have old versions deployed.

There is also some cleanup, moving things around and removing the keys endpoint from the controller.

refs #1744

### CODE REVIEW

File: controller/utils/utils.go [hunk 1]
```diff
@@ -71,3 +78,28 @@ type HostClient interface {
...
+var AppNamePattern = regexp.MustCompile(`^[a-z\d]+(-[a-z\d]+)*$`)
+
+func ParseBasicAuth(h http.Header) (username, password string, err error) {
+	s := strings.SplitN(h.Get("Authorization"), " ", 2)
+
+	if len(s) != 2 {
+		return "", "", errors.New("failed to parse authentication string ")
+	}
+	if s[0] != "Basic" {
+		return "", "", fmt.Errorf("authorization scheme is %v, not Basic ", s[0])
```
Comment (lmars) [inline controller/utils/utils.go]: Drop the trailing whitespace from this and the error above.

File: gitreceive/server.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+
+func (h *gitHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var g gitService
+
+	if r.URL.Path == status.Path {
+		status.HealthyHandler.ServeHTTP(w, r)
+		return
+	}
+
+	log.Printf("%s %q", r.Method, r.URL)
```
Comment (lmars) [inline gitreceive/server.go]: Should we wrap the handler in `httphelper.NewRequestLogger` to standardise our logging? Or at least use log15?

Comment (titanous) [inline gitreceive/server.go]: Yeah, I'll fix up the logging, this was derived from an existing codebase.

File: gitreceive/server.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+	// The client request body may have been gzipped.
+	if r.Header.Get("Content-Encoding") == "gzip" {
+		body, err = gzip.NewReader(r.Body)
+		if err != nil {
+			fail500(w, "handlePostRPC", err)
+			return
+		}
+	} else {
+		body = r.Body
```
Comment (lmars) [inline gitreceive/server.go]: Could set this above and avoid the `else`.

File: gitreceive/server.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+		return
+	}
+	defer cleanUpProcessGroup(cmd) // Ensure brute force subprocess clean-up
+
+	// Write the client request body to Git's standard input
+	if _, err := io.Copy(stdin, body); err != nil {
+		fail500(w, "handlePostRPC write to subprocess", err)
+		return
+	}
+	stdin.Close()
```
Comment (lmars) [inline gitreceive/server.go]: This close was deferred above, do we need to do it explicitly here?

File: gitreceive/server.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+	w.WriteHeader(200) // Don't bother with HTTP 500 from this point on, just return
+	if err := pktLine(w, fmt.Sprintf("# service=%s\n", rpc)); err != nil {
+		logContext("handleGetInfoRefs response", err)
+		return
+	}
+	if err := pktFlush(w); err != nil {
+		logContext("handleGetInfoRefs response", err)
+		return
+	}
+	if _, err := io.Copy(w, pipe); err != nil {
```
Comment (lmars) [inline gitreceive/server.go]: Should this use a `newFlushWriter` like `handlePostRPC`?

File: gitreceive/server.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+	}
+
+	process := cmd.Process
+	if process != nil && process.Pid > 0 {
+		// Send SIGTERM to the process group of cmd
+		syscall.Kill(-process.Pid, syscall.SIGTERM)
+	}
+
+	// reap our child process
+	cmd.Wait()
```
Comment (lmars) [inline gitreceive/server.go]: Do we need to explicitly wait for this (i.e. should we do it in a goroutine?)

**Comment by lmars** [comment]
Looks great! LGTM pending comments.

File: gitreceive/server.go (hunk 5 above)
Comment (titanous) [inline gitreceive/server.go]: It doesn't need it, there is no build output from this endpoint.

**Comment by titanous** [comment]
Comments addressed.

**Pull request merged**